### PR TITLE
fix(encrypt): refuse a filename dotenvx can't turn into a valid key (prevents secret lockout)

### PR DESCRIPTION
**Pull request by jainal09**
## What

P2 of the #443 re-audit (#23): `encrypt` on a filename with a space or non-ASCII char reports `[OK]` / exit 0 but leaves the file **permanently undecryptable** — the original plaintext is destroyed and the secret is locked out for good.

## Why

dotenvx derives the `DOTENV_PRIVATE_KEY_<SLUG>` variable name from the **filename**. `'my secrets.env'` → `DOTENV_PRIVATE_KEY_MY SECRETS…` — a space in an env-var name is invalid, so decrypt can never find the key. The existing plaintext-survival guard can't catch it (the value *is* encrypted; only the key name is unusable), and `encrypt --check` even reports "100% encrypted".

## Reproduction (real CLI)

```
printf 'SECRET=keepme\n' > 'my secrets.env'
envdrift encrypt 'my secrets.env'   # BEFORE: [OK]/exit 0, plaintext DESTROYED
envdrift decrypt 'my secrets.env'   # BEFORE: [ERROR] MISSING_PRIVATE_KEY, exit 1 — locked out
# AFTER: encrypt -> [ERROR] Refusing to encrypt … (exit 1), file untouched
```

## Fix

A pre-flight guard in the dotenvx backend (before the destructive call, and before `is_installed`) refuses to encrypt when the filename is not `[A-Za-z0-9._-]+`. Those are the names that round-trip safely (`postgresql.env`, `svc.env.docker`, `app-local.env` still encrypt); a space/unicode name now gets a clear `[ERROR]` to rename it. The original file is left byte-for-byte untouched.

## Tests (dogfood, test-first)

- **New** `test_encrypt_filename_with_space_refused_preserves_plaintext` — **red** before the fix (encrypt succeeded and destroyed the file), green after (refused, plaintext byte-identical).
- **New** `test_encrypt_unicode_filename_refused` — same for `café.env`.
- Existing custom-filename round-trip + the full 89-test encryption-backends module still pass.

Part of the #443 backlog (P2 of 9). See the [re-audit](https://github.com/jainal09/envdrift/issues/443#issuecomment-4663943129).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

Adds a pre-flight filename-validity guard to `DotenvxEncryptionBackend.encrypt` that refuses to proceed when the filename contains characters dotenvx cannot slugify into a valid env-var key name. Without this guard, encrypting `my secrets.env` exits 0 but leaves the file permanently undecryptable.

- **`src/envdrift/encryption/dotenvx.py`**: Inserts a `re.fullmatch(r"[A-Za-z0-9._-]+", env_file.name)` check after the empty-file guard; returns a descriptive `EncryptionResult(success=False, …)` and leaves the file byte-for-byte untouched when the filename fails.
- **`tests/unit/test_encryption_backends.py`**: Adds two regression tests (space in filename, non-ASCII filename) that assert `success is False`, pin the failure to the filename guard via `"undecryptable" in result.message.lower()`, and verify the original plaintext is preserved.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the change is a pre-flight guard that only adds a new early-return path; existing filenames that already round-trip correctly are unaffected.

The fix is a narrow, well-tested addition: a single `re.fullmatch` check returns a descriptive failure before any destructive I/O. The regex is correct, `re` is already imported, the character class is well-formed, and the two new tests pin the guard to the right code path via message-substring assertion and byte-exact file verification. No existing behaviour is altered.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/envdrift/encryption/dotenvx.py | Adds a correctly-anchored filename guard using `re.fullmatch`; `re` is already imported, the regex character class is well-formed, and the return path leaves the file untouched. |
| tests/unit/test_encryption_backends.py | Two new regression tests cover space and unicode filenames; both assert on `success`, message substring, and byte-exact file content, adequately pinning the new guard. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[encrypt called] --> B{File exists?}
    B -- No --> C[Return failure: file not found]
    B -- Yes --> D{file_has_assignment?}
    D -- No --> E[Return failure: nothing to encrypt]
    D -- Yes --> F{re.fullmatch\nA-Za-z0-9._- filename?}
    F -- No --> G[Return failure: invalid filename\nfile left untouched]
    F -- Yes --> H{is_installed?}
    H -- No --> I[Raise EncryptionNotFoundError]
    H -- Yes --> J[dotenvx encrypt]
    J --> K{has_plaintext_secret_value?}
    K -- Yes --> L[Return failure: encryption didn't take effect]
    K -- No --> M[Return success]
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `src/envdrift/encryption/dotenvx.py`, line 107-114 ([link](https://github.com/jainal09/envdrift/blob/b0f733b0eb44c953cc5ad71dffb829f67f611805/src/envdrift/encryption/dotenvx.py#L107-L114)) 

   <a href="#">P2</a> **Filename validity is a pure string check but runs after disk I/O**

   `file_has_assignment` opens and reads the file before the filename is even inspected. Since the filename check is zero-cost, moving it immediately after `env_file = Path(env_file)` (before the exists check or any I/O) makes the common bad-name path faster and removes any accidental dependency on file content. The ordering doesn't cause a correctness bug today, but the current placement is surprising given the comment says "before the destructive call, and before `is_installed`".
</details>

<sub>Reviews (2): Last reviewed commit: ["test(encrypt): assert the filename-guard..."](https://github.com/jainal09/envdrift/commit/cc825ad1cb1638f4182109de5b59ce9bf1370cfe) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36586669)</sub>

### CODE REVIEW

File: tests/unit/test_encryption_backends.py [hunk 1]
```diff
@@ -283,6 +283,36 @@ def test_encrypt_file_not_found(self, tmp_path):
...
+    def test_encrypt_unicode_filename_refused(self, tmp_path):
+        """#23: a non-ASCII filename also yields an invalid dotenvx key name."""
+        env_file = tmp_path / "café.env"
+        env_file.write_text("SECRET=keepme\n")
+
+        backend = DotenvxEncryptionBackend()
+        result = backend.encrypt(env_file)
+
+        assert result.success is False
+        assert env_file.read_text() == "SECRET=keepme\n"
```
[the remark this answers is not in the document]
Comment (jainal09) [reply tests/unit/test_encryption_backends.py:317]: Good point — adopted in cc825ad: both tests now `assert "undecryptable" in result.message.lower()`, pinning the refusal to the filename guard so a regression in another guard (e.g. empty-file) can no longer keep them green.

File: src/envdrift/encryption/dotenvx.py [hunk 2]
```diff
@@ -135,6 +135,27 @@ def encrypt(
...
+                success=False,
+                message=(
+                    f"Refusing to encrypt {env_file}: its filename contains "
+                    "characters dotenvx cannot turn into a valid key name, which "
+                    "would leave the file permanently undecryptable. Rename it to "
+                    "use only letters, digits, '.', '-' and '_'."
+                ),
+                file_path=env_file,
+            )
+
```
[the remark this answers is not in the document]
Comment (jainal09) [reply src/envdrift/encryption/dotenvx.py:158]: Advisory (non-required). The added pre-flight guard is a flat early-return `if`/`return` with no nesting or new branches in the existing logic — it prevents permanent secret loss, which outweighs the +1 to the method length metric. Refactoring `encrypt` is out of scope for this security fix.

**Comment by jainal09** [comment]
🤖 **envdrift review-bot** — automated re-review of commit cc825ad


The new filename-validity guard fixes the backend path but the data-loss scope of this PR is still open: the partial-encryption push/lock path bypasses the guarded backend entirely.

**Findings (1):**
- [medium] Secret-lockout guard is backend-only; partial-encryption (push/pull --secrets-only, lock) bypasses it via `DotenvxWrapper.encrypt` and can still permanently lock out a space/unicode-named file — `src/envdrift/core/partial_encryption.py` (not anchored inline — vulnerable call sites are pre-existing code outside this PR's diff)

### Not anchored inline

**[medium · bug] Secret-lockout guard is backend-only; partial-encryption (push/pull --secrets-only, lock) bypasses it via DotenvxWrapper.encrypt and can still permanently lock out a space/unicode-named file**
`src/envdrift/core/partial_encryption.py`


The new filename guard added in this PR lives only in `DotenvxEncryptionBackend.encrypt` (src/envdrift/encryption/dotenvx.py). But the partial-encryption code path does **not** go through that backend — it calls the low-level integration wrapper `DotenvxWrapper.encrypt()` directly:

- `encrypt_secret_file` (partial_encryption.py ~line 519): `dotenvx.encrypt(secret_file)`
- `_encrypt_secrets_only_file` invoked from `push_secrets_only` (partial_encryption.py ~line 704): `dotenvx.encrypt(file)`, where `file` comes from `secrets_dir.glob(env_config.pattern)` — i.e. **arbitrary on-disk filenames matching a user-configured glob**.

`DotenvxWrapper.encrypt` only runs `_validate_filename`, and that check (PROBLEMATIC_FILENAME_PATTERNS = [`^\.env\.local$`]) is Windows-`.env.local`-only and is additionally gated behind `if platform.system() == 'Windows'`. It does **not** reject spaces or non-ASCII characters. So a user who has e.g. `my secret.env` or `café.env.secret` in their `secrets_dir` and runs `envdrift push`/`lock` (secrets-only) hits the exact same permanent-lockout / data-loss bug this PR set out to fix: the value encrypts, dotenvx exits 0, the original plaintext is destroyed, and the derived `DOTENV_PRIVATE_KEY_<bad slug>` name is unusable so decrypt can never recover it.

I empirically confirmed in /tmp with dotenvx 1.66.0 that `my secrets.env`, `café.env`, `app+prod.env` etc. all encrypt to `roundtrip=0` (un-decryptable), while the wrapper performs no slug check.

Impact: the fix is incomplete for the issue's data-loss scope — the most automation-heavy path (push/lock over a globbed secrets dir, where users are least likely to be watching filenames) is still vulnerable. Direction: move/share the slug-validity guard so it also covers `DotenvxWrapper.encrypt` (or have the partial-encryption callers route through the guarded backend), and add a regression test that drives `push_secrets_only`/`encrypt_secret_file` with a space/unicode-named secret file and asserts it refuses with plaintext preserved.
